include/gridtools/topo.hpp

```diff
@@ -0,0 +1,191 @@
+                meta::transform<front_push_f<entity<T, U>>::template apply, rhs_t>>;
+        };
+
+        template <class>
+        struct links;
+
+        template <location T, location U>
+        struct links<entity<T, U>> {
+            using type = meta::list<entity<T, U>>;
+        };
+
+        template <location T, location U, location... Ts>
+        struct links<entity<T, U, Ts...>> {
+            using type = meta::push_back<typename links<entity<U, Ts...>>::type, entity<T, U>>;
+        };
+
+        template <class>
+        struct reverse;
+
+        template <location... Locations>
+        struct reverse<entity<Locations...>> {
+            using type = decltype((entity<Locations>(), ...));
+        };
+
+        template <class>
+        struct to;
+
+        template <location T, location U, location... Ts>
+        struct to<entity<T, U, Ts...>> {
+            using type = entity<T>;
+        };
+
+        template <class>
+        struct is_location : std::false_type {};
+
+        template <location Location>
+        struct is_location<entity<Location>> : std::true_type {};
+
+        template <class, class = void>
+        struct is_link : std::false_type {};
+
+        template <location U, location T>
+        struct is_link<entity<U, T>, std::enable_if_t<U != T && (U == location::edge || T == location::edge)>>
+            : std::true_type {};
+
+        template <class T, class = void>
+        struct is_chain : is_link<T> {};
+
+        template <location U, location T, location... Ts>
+        struct is_chain<entity<U, T, Ts...>, std::enable_if_t<U != T && is_chain<entity<T, Ts...>>::value>>
+            : std::true_type {};
+    } // namespace _impl
+
+    // topological locations
+    using vertex = _impl::entity<_impl::location::vertex>;
+    using edge = _impl::entity<_impl::location::edge>;
+    using cell = _impl::entity<_impl::location::cell>;
+
+    using locations = meta::list<vertex, edge, cell>;
+
+    //  Neighbor chains are produced with the following syntax:
+    //  `vertex::edge:: ... ::vertex::cell`
+    //  The meaning is:
+    //    the `cell` neighbors of the `vertices` that are neighbors of ... `edges` of the given `vertex`.
+    //  The links in the chain from and to the same location type are disallowed.
+    //  I.e. `cell::cell` is invalid.
+    //
+    // Internally chains are canonicalized during creation:
+    // `vertex::cell` links of the chain are replaced by equivalent `vertex::edge::cell`.
+    // `cell::vertex` links are replaced by `cell::edge::cell`.
+    // i.e.  static_assert(std::is_same_v<vertex::cell::vertex, vertex::edge::cell::edge::vertex>);
+
+    // link is the chain with two locations where one of them is `edge`.
+
+    using _impl::is_chain;
+    using _impl::is_link;
+    using _impl::is_location;
+
+    template <class Chain>
+    using reverse = typename _impl::reverse<Chain>::type;
+
+    template <class Chain>
+    using to = typename _impl::to<Chain>::type;
```

Comment:
`to` and `from` are not intuitive for me. From the names I would expect that they construct a chain. Maybe `begin`/ `end` or `first`/`last`?

Reply:
done